packages/server-wallet/src/protocols/process-ledger-queue.ts

```diff
@@ -0,0 +1,198 @@
+import {compose, map, filter} from 'lodash/fp';
+import _ from 'lodash';
+import {
+  allocateToTarget,
+  AllocationItem,
+  BN,
+  checkThat,
+  isSimpleAllocation,
+  Outcome,
+  SignedStateWithHash,
+  SimpleAllocation,
+} from '@statechannels/wallet-core';
+import {Transaction} from 'knex';
+
+import {Store} from '../wallet/store';
+import {Bytes32} from '../type-aliases';
+import {LedgerRequestType} from '../models/ledger-requests';
+
+import {Protocol, ProtocolResult, ChannelState} from './state';
+import {
+  MarkLedgerFundingRequestsAsComplete,
+  noAction,
+  ProtocolAction,
+  SignState,
+  signState,
+} from './actions';
+
+export type ProtocolState = {
+  fundingChannel: ChannelState;
+  channelsPendingRequest: ChannelState[];
+};
+
+const allocationItemComparator = (a: AllocationItem, b: AllocationItem): boolean =>
+  a.destination === b.destination && BN.eq(a.amount, b.amount);
+
+const foldAllocateToTarget = (
+  original: SimpleAllocation,
+  allocationTargets: ChannelState[]
+): SimpleAllocation =>
+  // This could fail at some point if there is no longer space to fund stuff
+  // TODO: Handle that case
+  // } catch (e) {
+  //   if (e.toString() === 'Insufficient funds in fundingChannel channel')
+  // }
+  allocationTargets.reduce(
+    (outcome, {channelId, supported}) =>
+      checkThat(
+        allocateToTarget(
+          outcome,
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          checkThat(supported!.outcome, isSimpleAllocation).allocationItems,
+          channelId
+        ),
+        isSimpleAllocation
+      ),
+    original
+  );
+
+const mergedOutcome = (a: SimpleAllocation, b: SimpleAllocation): SimpleAllocation => ({
+  type: 'SimpleAllocation',
+  assetHolderAddress: b.assetHolderAddress,
+  allocationItems: _.intersectionWith(
+    a.allocationItems,
+    b.allocationItems,
+    allocationItemComparator
+  ),
+});
+
+const computeNewOutcome = ({
+  fundingChannel: {
+    supported,
+    latestNotSignedByMe,
+    latestSignedByMe,
+    channelId,
+    participants: {length: n},
+  },
+  channelsPendingRequest,
+}: ProtocolState): SignState | false => {
+  // Sanity-checks
+  if (!supported) return false;
+  if (!latestSignedByMe) return false;
+
+  const myLatestOutcome = checkThat(latestSignedByMe.outcome, isSimpleAllocation);
+  const supportedOutcome = checkThat(supported.outcome, isSimpleAllocation);
+
+  // Nothing left to do, no actions to take
+  if (channelsPendingRequest.length === 0) return false;
+
+  // TODO: Sort should be somewhere else
+  channelsPendingRequest = channelsPendingRequest.sort(
+    (a, b) => a.latest.channelNonce - b.latest.channelNonce
+  );
+
+  const receivedOriginal =
+    latestNotSignedByMe && latestNotSignedByMe.turnNum === supported.turnNum + 2;
+  const receivedMerged =
+    latestNotSignedByMe && latestNotSignedByMe.turnNum === supported.turnNum + 2 * n;
```

Comment:
request: Can this do validations previously so that we know that `latestNotSignedByMe` is defined?

Then it can look like
```
  const receivedOriginal = latestNotSignedByMe.turnNum === supported.turnNum + 2;
  const receivedMerged = latestNotSignedByMe.turnNum === supported.turnNum + 2 * n;
  const sentOriginal = latestSignedByMe.turnNum === supported.turnNum + n;
  const sentMerged = latestSignedByMe.turnNum === supported.turnNum + 2 * n;
```

Reply:
It's even more readable in GitHub's editor 😄 

<img width="601" alt="Screen Shot 2020-10-28 at 7 10 41 PM" src="https://user-images.githubusercontent.com/8432675/97513215-46743b80-1951-11eb-987e-2a908f0bd936.png">

Reply:
I still don't understand these booleans, after a few minutes thought :(

Reply:
I think that `computeNewOutcome` needs a rough description of the algorithm in a comment, before its implementation.

Reply:
I think I need to check `latestNotSignedByMe` not being `undefined` because it could be `undefined` if I have signed everything so far.

> I think that computeNewOutcome needs a rough description of the algorithm in a comment, before its implementation.

Will add something like this -- but also see #2761 which I'd prefer to do a writeup on. Also see https://github.com/statechannels/statechannels/pull/2699#issue-504910127.

Reply:
bbb636b